src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs

```diff
@@ -258,180 +255,49 @@ private unsafe IAsyncResult BeginWaitForConnection(AsyncCallback callback, Objec
 
             // Create and store async stream class library specific data in the 
             // async result
-            PipeAsyncResult asyncResult = new PipeAsyncResult();
-            asyncResult._threadPoolBinding = _threadPoolBinding;
-            asyncResult._userCallback = callback;
-            asyncResult._userStateObject = state;
+            var completionSource = new ConnectionCompletionSource(
+                _threadPoolBinding,
+                cancellationToken,
+                this,
+                onSuccess: thisRef => thisRef.State = PipeState.Connected);
```

Comment:
We only ever construct this ConnectionCompletionSource here, and only ever with this specific callback.  Could we reduce the size of ConnectionCompletionSource by avoiding the need for this delegate and just baking the ```State = PipeState.Connected``` logic into it?